# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [2]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! y

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [3]:
import problem_unittests as tests

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    vocab_to_int = {char:i for (i,char) in enumerate(set(text))}
    int_to_vocab = {v:k for (k,v) in vocab_to_int.items()}
    # return tuple
    return (vocab_to_int,int_to_vocab)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [4]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
    tokens = {
        
        '.': '||period||',
        ',': '||comma||',
        '"': '||quotation_mark||',
        ';': '||semicolon||',
        '!': '||exclamation_mark||',
        '?': '||question_mark||',
        '(': '||left_parentheses||',
        ')': '||right_Parentheses||',
        '-': '||dash||',
        '\n': '||return||'
 }
        
    return tokens

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to lok at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [8]:
from torch.utils.data import TensorDataset, DataLoader


def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # TODO: Implement function
    features = []
    target = []
    # return a dataloader
    for i in range(len(words) - sequence_length):
        features.append(words[i:i+sequence_length])
        target.append(words[i+sequence_length])
    data = TensorDataset(torch.tensor(features),torch.tensor(target))
    data_loader = torch.utils.data.DataLoader(data,shuffle=True,batch_size=batch_size,num_workers=8)
    return data_loader

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own


### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [9]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[ 26,  27,  28,  29,  30],
        [ 29,  30,  31,  32,  33],
        [  7,   8,   9,  10,  11],
        [  8,   9,  10,  11,  12],
        [ 15,  16,  17,  18,  19],
        [ 28,  29,  30,  31,  32],
        [ 31,  32,  33,  34,  35],
        [ 33,  34,  35,  36,  37],
        [ 17,  18,  19,  20,  21],
        [  1,   2,   3,   4,   5]])

torch.Size([10])
tensor([ 31,  34,  12,  13,  20,  33,  36,  38,  22,   6])


In [13]:
!curl -o workspace_utils.py https://s3.amazonaws.com/video.udacity-data.com/topher/2018/May/5b0dea96_workspace-utils/workspace-utils.py

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1540  100  1540    0     0   8782      0 --:--:-- --:--:-- --:--:-- 10065


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [14]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        # TODO: Implement function
        
        # set class variables
        
        # define model layers
        self.input_size = vocab_size
        self.hidden_dim = hidden_dim
        self.output_size = output_size
        self.n_layers = n_layers
        self.dropout = dropout
        self.embed = nn.Embedding(vocab_size,embedding_dim)
        self.lstm = nn.LSTM(embedding_dim,hidden_dim,n_layers,dropout=dropout,batch_first = True)
        self.fc = nn.Linear(hidden_dim,output_size)
    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # TODO: Implement function   
        batch_size = nn_input.size(0)
        embed_output = self.embed(nn_input)
        lstm_out,hidden = self.lstm(embed_output,hidden)
        lstm_out = lstm_out.contiguous().view(-1,self.hidden_dim)
        out = self.fc(lstm_out)
        out = out.view(batch_size,-1,self.output_size)
        out=out[:,-1]
        # return one batch of output word scores and the hidden state
        return out,hidden
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        weight =next(self.parameters()).data
        # Implement function
        
        # initialize hidden state with zero weights, and move to GPU if available
        if torch.cuda.is_available():
            hidden = (weight.new(self.n_layers,batch_size,self.hidden_dim).zero_().cuda(),
                     weight.new(self.n_layers,batch_size,self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers,batch_size,self.hidden_dim).zero_(),
                     weight.new(self.n_layers,batch_size,self.hidden_dim).zero_())
        return hidden
    

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [15]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param decoder: The PyTorch Module that holds the neural network
    :param decoder_optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # TODO: Implement Function
    if torch.cuda.is_available():
        inp,target = inp.cuda(),target.cuda()
    # move data to GPU, if available
    hidden = tuple([i.data for i in hidden])
    rnn.zero_grad()
    out,hidden = rnn(inp,hidden)
    loss = criterion(out,target)
    loss.backward()
    nn.utils.clip_grad_norm_(rnn.parameters(),5)
    
    # perform backpropagation and optimization
    optimizer.step()
    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), hidden

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [24]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from tqdm import tqdm
from workspace_utils import active_session


def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in tqdm(range(1, n_epochs + 1)):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            if batch_i % show_every_n_batches == 0:
                print('Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
                    epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of uniqe tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [20]:
# Data params
# Sequence Length
sequence_length =  32 # of words in a sequence
# Batch Size
batch_size = 128

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [21]:
# Training parameters
# Number of Epochs
num_epochs = 30
# Learning Rate
learning_rate = 0.0005

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 256
# Hidden Dimension
hidden_dim = 256
# Number of RNN Layers
n_layers = 3

# Show stats for every n number of batches
show_every_n_batches = 500

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
with active_session():
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')


  0%|          | 0/30 [00:00<?, ?it/s]

Training for 30 epoch(s)...


Epoch:    1/30    Loss: 5.953537424087524

Epoch:    1/30    Loss: 5.199837107658386

Epoch:    1/30    Loss: 4.9583313245773315

Epoch:    1/30    Loss: 4.74512848329544

Epoch:    1/30    Loss: 4.69165274810791

Epoch:    1/30    Loss: 4.582411173343658

Epoch:    1/30    Loss: 4.530147205352783

Epoch:    1/30    Loss: 4.481191950798035

Epoch:    1/30    Loss: 4.459470741271972

Epoch:    1/30    Loss: 4.40465577507019

Epoch:    1/30    Loss: 4.383221363067627

Epoch:    1/30    Loss: 4.364323173046112

Epoch:    1/30    Loss: 4.354030261039734



  3%|▎         | 1/30 [14:13<6:52:17, 853.01s/it]

Epoch:    2/30    Loss: 4.242714292370744

Epoch:    2/30    Loss: 4.17018239402771

Epoch:    2/30    Loss: 4.153783295631409

Epoch:    2/30    Loss: 4.136279977798462

Epoch:    2/30    Loss: 4.11864275932312

Epoch:    2/30    Loss: 4.121788383960724

Epoch:    2/30    Loss: 4.1021428713798525

Epoch:    2/30    Loss: 4.100882353782654

Epoch:    2/30    Loss: 4.107127541065216

Epoch:    2/30    Loss: 4.096107387065888

Epoch:    2/30    Loss: 4.076151418685913

Epoch:    2/30    Loss: 4.0540254435539245

Epoch:    2/30    Loss: 4.056558194637298



  7%|▋         | 2/30 [28:52<6:41:44, 860.88s/it]

Epoch:    3/30    Loss: 3.9954907958106483

Epoch:    3/30    Loss: 3.9425565857887266

Epoch:    3/30    Loss: 3.9213106245994567

Epoch:    3/30    Loss: 3.9167117466926573

Epoch:    3/30    Loss: 3.90386080121994

Epoch:    3/30    Loss: 3.92667116689682

Epoch:    3/30    Loss: 3.913003935813904

Epoch:    3/30    Loss: 3.901222092628479

Epoch:    3/30    Loss: 3.9125028367042542

Epoch:    3/30    Loss: 3.8992103333473205

Epoch:    3/30    Loss: 3.9193982486724854

Epoch:    3/30    Loss: 3.9237143025398256

Epoch:    3/30    Loss: 3.931635901927948



 10%|█         | 3/30 [43:05<6:26:22, 858.63s/it]

Epoch:    4/30    Loss: 3.852229357011793

Epoch:    4/30    Loss: 3.8050891857147215

Epoch:    4/30    Loss: 3.8187318286895753

Epoch:    4/30    Loss: 3.7718096261024474

Epoch:    4/30    Loss: 3.770264241218567

Epoch:    4/30    Loss: 3.7955603652000427

Epoch:    4/30    Loss: 3.784201994895935

Epoch:    4/30    Loss: 3.788107623577118

Epoch:    4/30    Loss: 3.797271430015564

Epoch:    4/30    Loss: 3.797329345226288

Epoch:    4/30    Loss: 3.806253830909729

Epoch:    4/30    Loss: 3.812298598766327

Epoch:    4/30    Loss: 3.802727689743042



 13%|█▎        | 4/30 [57:43<6:14:36, 864.50s/it]

Epoch:    5/30    Loss: 3.7384865434054126

Epoch:    5/30    Loss: 3.6695009703636168

Epoch:    5/30    Loss: 3.6874551610946655

Epoch:    5/30    Loss: 3.685362955093384

Epoch:    5/30    Loss: 3.6725590643882753

Epoch:    5/30    Loss: 3.7100112695693968

Epoch:    5/30    Loss: 3.710571357727051

Epoch:    5/30    Loss: 3.697425477027893

Epoch:    5/30    Loss: 3.719692766189575

Epoch:    5/30    Loss: 3.698128664970398

Epoch:    5/30    Loss: 3.726002028465271

Epoch:    5/30    Loss: 3.727807321548462

Epoch:    5/30    Loss: 3.7204896397590637



 17%|█▋        | 5/30 [1:11:57<5:58:52, 861.32s/it]

Epoch:    6/30    Loss: 3.6698276477455476

Epoch:    6/30    Loss: 3.5646275010108948

Epoch:    6/30    Loss: 3.590862710952759

Epoch:    6/30    Loss: 3.6294324951171877

Epoch:    6/30    Loss: 3.6104354028701784

Epoch:    6/30    Loss: 3.610529538631439

Epoch:    6/30    Loss: 3.6238822660446166

Epoch:    6/30    Loss: 3.626958607673645

Epoch:    6/30    Loss: 3.621484573841095

Epoch:    6/30    Loss: 3.6226193056106566

Epoch:    6/30    Loss: 3.6541114459037782

Epoch:    6/30    Loss: 3.658002100467682

Epoch:    6/30    Loss: 3.667737215042114



 20%|██        | 6/30 [1:26:37<5:46:42, 866.77s/it]

Epoch:    7/30    Loss: 3.5794732693671194

Epoch:    7/30    Loss: 3.5463682899475097

Epoch:    7/30    Loss: 3.523266100883484

Epoch:    7/30    Loss: 3.5411159801483154

Epoch:    7/30    Loss: 3.552560224056244

Epoch:    7/30    Loss: 3.55966642665863

Epoch:    7/30    Loss: 3.560334179401398

Epoch:    7/30    Loss: 3.536092694282532

Epoch:    7/30    Loss: 3.5834116611480713

Epoch:    7/30    Loss: 3.575403305530548

Epoch:    7/30    Loss: 3.587728600502014

Epoch:    7/30    Loss: 3.5878522720336914

Epoch:    7/30    Loss: 3.583400604248047



 23%|██▎       | 7/30 [1:40:51<5:30:46, 862.88s/it]

Epoch:    8/30    Loss: 3.5337454883921873

Epoch:    8/30    Loss: 3.462372189998627

Epoch:    8/30    Loss: 3.472190429210663

Epoch:    8/30    Loss: 3.473314881324768

Epoch:    8/30    Loss: 3.478488851070404

Epoch:    8/30    Loss: 3.4994991707801817

Epoch:    8/30    Loss: 3.4943327078819273

Epoch:    8/30    Loss: 3.500461639881134

Epoch:    8/30    Loss: 3.5257844977378845

Epoch:    8/30    Loss: 3.5251084384918214

Epoch:    8/30    Loss: 3.5280225410461425

Epoch:    8/30    Loss: 3.527146436691284

Epoch:    8/30    Loss: 3.5578390731811522



 27%|██▋       | 8/30 [1:55:31<5:18:16, 868.03s/it]

Epoch:    9/30    Loss: 3.487668460367634

Epoch:    9/30    Loss: 3.406313817501068

Epoch:    9/30    Loss: 3.4371124744415282

Epoch:    9/30    Loss: 3.4240428786277772

Epoch:    9/30    Loss: 3.444076099395752

Epoch:    9/30    Loss: 3.4611661972999572

Epoch:    9/30    Loss: 3.454951357841492

Epoch:    9/30    Loss: 3.4765474400520326

Epoch:    9/30    Loss: 3.4851556677818296

Epoch:    9/30    Loss: 3.4760525064468384

Epoch:    9/30    Loss: 3.4744198198318483

Epoch:    9/30    Loss: 3.4822200961112975

Epoch:    9/30    Loss: 3.5000873823165892



 30%|███       | 9/30 [2:09:45<5:02:22, 863.94s/it]

Epoch:   10/30    Loss: 3.4374954506093625

Epoch:   10/30    Loss: 3.4045449776649477

Epoch:   10/30    Loss: 3.3810041561126707

Epoch:   10/30    Loss: 3.390115671157837

Epoch:   10/30    Loss: 3.4120547909736634

Epoch:   10/30    Loss: 3.393798038959503

Epoch:   10/30    Loss: 3.4096869320869447

Epoch:   10/30    Loss: 3.4294102997779845

Epoch:   10/30    Loss: 3.4267267990112305

Epoch:   10/30    Loss: 3.446013090610504

Epoch:   10/30    Loss: 3.4325038628578186

Epoch:   10/30    Loss: 3.456198446750641

Epoch:   10/30    Loss: 3.4622639031410216



 33%|███▎      | 10/30 [2:24:25<4:49:34, 868.72s/it]

Epoch:   11/30    Loss: 3.3960902594929507

Epoch:   11/30    Loss: 3.343825267314911

Epoch:   11/30    Loss: 3.3618866314888

Epoch:   11/30    Loss: 3.350920918941498

Epoch:   11/30    Loss: 3.3434061579704286

Epoch:   11/30    Loss: 3.385610674858093

Epoch:   11/30    Loss: 3.368828616142273

Epoch:   11/30    Loss: 3.369482639312744

Epoch:   11/30    Loss: 3.4016582674980165

Epoch:   11/30    Loss: 3.4267741446495057

Epoch:   11/30    Loss: 3.411659372806549

Epoch:   11/30    Loss: 3.4117858967781065

Epoch:   11/30    Loss: 3.4255009684562685



 37%|███▋      | 11/30 [2:38:39<4:33:43, 864.38s/it]

Epoch:   12/30    Loss: 3.3674233050784337

Epoch:   12/30    Loss: 3.2950123624801635

Epoch:   12/30    Loss: 3.3076448760032653

Epoch:   12/30    Loss: 3.3157617292404176

Epoch:   12/30    Loss: 3.336520754814148

Epoch:   12/30    Loss: 3.3493055982589723

Epoch:   12/30    Loss: 3.344413845062256

Epoch:   12/30    Loss: 3.349934257030487

Epoch:   12/30    Loss: 3.3554476027488707

Epoch:   12/30    Loss: 3.352575839519501

Epoch:   12/30    Loss: 3.37268248128891

Epoch:   12/30    Loss: 3.3813384985923767

Epoch:   12/30    Loss: 3.3901339111328124



 40%|████      | 12/30 [2:53:18<4:20:39, 868.85s/it]

Epoch:   13/30    Loss: 3.338361711078876

Epoch:   13/30    Loss: 3.265438148975372

Epoch:   13/30    Loss: 3.272463391780853

Epoch:   13/30    Loss: 3.29286075592041

Epoch:   13/30    Loss: 3.305274182796478

Epoch:   13/30    Loss: 3.2925346326828

Epoch:   13/30    Loss: 3.310701020717621

Epoch:   13/30    Loss: 3.3146590342521667

Epoch:   13/30    Loss: 3.3386180024147034

Epoch:   13/30    Loss: 3.326850869655609

Epoch:   13/30    Loss: 3.350093596935272

Epoch:   13/30    Loss: 3.354546105861664

Epoch:   13/30    Loss: 3.3497613973617555



 43%|████▎     | 13/30 [3:07:33<4:04:57, 864.55s/it]

Epoch:   14/30    Loss: 3.3114790611345826

Epoch:   14/30    Loss: 3.23788862657547

Epoch:   14/30    Loss: 3.2377460107803344

Epoch:   14/30    Loss: 3.2540839133262636

Epoch:   14/30    Loss: 3.277641932964325

Epoch:   14/30    Loss: 3.288545524597168

Epoch:   14/30    Loss: 3.2927146983146667

Epoch:   14/30    Loss: 3.2934103660583496

Epoch:   14/30    Loss: 3.3209978399276734

Epoch:   14/30    Loss: 3.3249436206817626

Epoch:   14/30    Loss: 3.2995576982498167

Epoch:   14/30    Loss: 3.3324783368110658

Epoch:   14/30    Loss: 3.317405837059021



 47%|████▋     | 14/30 [3:22:13<3:51:47, 869.20s/it]

Epoch:   15/30    Loss: 3.2750222363093076

Epoch:   15/30    Loss: 3.207557194232941

Epoch:   15/30    Loss: 3.2126724519729613

Epoch:   15/30    Loss: 3.2419247670173643

Epoch:   15/30    Loss: 3.236738224029541

Epoch:   15/30    Loss: 3.257000823020935

Epoch:   15/30    Loss: 3.2663816637992857

Epoch:   15/30    Loss: 3.2867986664772033

Epoch:   15/30    Loss: 3.273063060760498

Epoch:   15/30    Loss: 3.2784780921936036

Epoch:   15/30    Loss: 3.297523545265198

Epoch:   15/30    Loss: 3.3063100519180297

Epoch:   15/30    Loss: 3.302955895423889



 50%|█████     | 15/30 [3:36:28<3:36:12, 864.82s/it]

Epoch:   16/30    Loss: 3.2405215239746283

Epoch:   16/30    Loss: 3.2116069416999817

Epoch:   16/30    Loss: 3.1991767807006837

Epoch:   16/30    Loss: 3.2126284914016723

Epoch:   16/30    Loss: 3.226695876121521

Epoch:   16/30    Loss: 3.2390148220062254

Epoch:   16/30    Loss: 3.233793833255768

Epoch:   16/30    Loss: 3.245987739562988

Epoch:   16/30    Loss: 3.2438506717681883

Epoch:   16/30    Loss: 3.274187171459198

Epoch:   16/30    Loss: 3.2636441173553465

Epoch:   16/30    Loss: 3.2733189687728883

Epoch:   16/30    Loss: 3.278289123058319



 53%|█████▎    | 16/30 [3:51:06<3:22:46, 869.02s/it]

Epoch:   17/30    Loss: 3.223654128695666

Epoch:   17/30    Loss: 3.174617355823517

Epoch:   17/30    Loss: 3.1818737053871153

Epoch:   17/30    Loss: 3.2046071085929873

Epoch:   17/30    Loss: 3.1813300862312315

Epoch:   17/30    Loss: 3.206632870197296

Epoch:   17/30    Loss: 3.209392825126648

Epoch:   17/30    Loss: 3.2129895806312563

Epoch:   17/30    Loss: 3.238340658664703

Epoch:   17/30    Loss: 3.246426833629608

Epoch:   17/30    Loss: 3.2456010794639587

Epoch:   17/30    Loss: 3.24565242433548

Epoch:   17/30    Loss: 3.2693911228179933



In [27]:
!ls

data				    helper.py		  __pycache__
dlnd_tv_script_generation.ipynb     preprocess.p	  trained_rnn.pt
dlnd_tv_script_generation-zh.ipynb  problem_unittests.py  workspace_utils.py


### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:** (Write answer, here)

I tried to have all parameters as multiples of 2. Primary hyper parameter that helped make the model converge faster was the learning rate - reduced it to 0.0005 instead of 0.001 and 0.005. Additionally tried with models with 512 embed and 512 hidden dims and it took almost 1.5 times more estimating around 12 hours and with my limited compute time wouldnt be able to complete. I kept batch size to 128 so that I will be able to try our more complex models. Understood from the lectures that going more than 3 layers wouldnt add much value. I got loss ~3.5 with 20 epochs with model with 512 embed layer, the convergence in 256 was slightly quicker. My computer turned off while running the model so had to reconnect. final loss after 31(30+1 below) is ~2.9. 

In [34]:
with active_session():
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, 1, show_every_n_batches)

  0%|          | 0/1 [00:00<?, ?it/s]

Training for 1 epoch(s)...
Epoch:    1/1     Loss: 2.962954560756683

Epoch:    1/1     Loss: 2.9477477502822875

Epoch:    1/1     Loss: 2.9609110922813415

Epoch:    1/1     Loss: 2.9890257859230043

Epoch:    1/1     Loss: 3.007437324523926

Epoch:    1/1     Loss: 2.9920752000808717

Epoch:    1/1     Loss: 3.0080960907936096

Epoch:    1/1     Loss: 3.009536564350128

Epoch:    1/1     Loss: 3.0125099220275877

Epoch:    1/1     Loss: 3.0309117560386656

Epoch:    1/1     Loss: 3.031012731075287

Epoch:    1/1     Loss: 3.057153111934662

Epoch:    1/1     Loss: 3.0476337571144105



100%|██████████| 1/1 [14:14<00:00, 854.13s/it]


---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [30]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [31]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        # the generated word becomes the next "current sequence" and the cycle can continue
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [35]:
# run the cell multiple times to get different results!
gen_length = 400 # modify the length to your preference
prime_word = 'jerry' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

jerry: over.

george:(pause) well, i was a very nervous electronics. i don't know if you can do that.

jerry: oh, no...

elaine: no, i don't have to do anything.

elaine: oh!

jerry: you got a date?

kramer: no, i don't want it.

jerry: i don't know, but i don't have a massage.

george: what are you talking about?

jerry: i think it's dutch. i was thinking about it, so i can't stand out of the way to the end of the night.

elaine: i don't know, maybe i was gonna get it out, i have no idea...

george: no, you can't believe that, you know, i can't do that.(kramer is speechless)

jerry:(to the phone) what are you doing here, huh?

george: oh, no, no, no, no. i got it, and i was thinking of the game. i mean, it's not a big thing i am.(to george) i have to tell you, i don't know.

jerry:(jokingly hits his hand) oh, i don't know.(kramer throws his head up and down for the window, and starts to move on).

elaine:(to jerry) what do you think?

frank: you know what i mean, i was in the shop wit

#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [36]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.